src/backend/vulkan/src/physical_device.rs

```diff
@@ -12,162 +12,673 @@ use hal::{
+                )
+                .shader_storage_buffer_array_dynamic_indexing(
+                    features.contains(Features::SHADER_STORAGE_BUFFER_ARRAY_DYNAMIC_INDEXING),
+                )
+                .shader_storage_image_array_dynamic_indexing(
+                    features.contains(Features::SHADER_STORAGE_IMAGE_ARRAY_DYNAMIC_INDEXING),
+                )
+                .shader_clip_distance(features.contains(Features::SHADER_CLIP_DISTANCE))
+                .shader_cull_distance(features.contains(Features::SHADER_CULL_DISTANCE))
+                .shader_float64(features.contains(Features::SHADER_FLOAT64))
+                .shader_int64(features.contains(Features::SHADER_INT64))
+                .shader_int16(features.contains(Features::SHADER_INT16))
+                .shader_resource_residency(features.contains(Features::SHADER_RESOURCE_RESIDENCY))
+                .shader_resource_min_lod(features.contains(Features::SHADER_RESOURCE_MIN_LOD))
+                .sparse_binding(features.contains(Features::SPARSE_BINDING))
+                .sparse_residency_buffer(features.contains(Features::SPARSE_RESIDENCY_BUFFER))
+                .sparse_residency_image2_d(features.contains(Features::SPARSE_RESIDENCY_IMAGE_2D))
+                .sparse_residency_image3_d(features.contains(Features::SPARSE_RESIDENCY_IMAGE_3D))
+                .sparse_residency2_samples(features.contains(Features::SPARSE_RESIDENCY_2_SAMPLES))
+                .sparse_residency4_samples(features.contains(Features::SPARSE_RESIDENCY_4_SAMPLES))
+                .sparse_residency8_samples(features.contains(Features::SPARSE_RESIDENCY_8_SAMPLES))
+                .sparse_residency16_samples(
+                    features.contains(Features::SPARSE_RESIDENCY_16_SAMPLES),
+                )
+                .sparse_residency_aliased(features.contains(Features::SPARSE_RESIDENCY_ALIASED))
+                .variable_multisample_rate(features.contains(Features::VARIABLE_MULTISAMPLE_RATE))
+                .inherited_queries(features.contains(Features::INHERITED_QUERIES))
+                .build(),
+            vulkan_1_2: if api_version >= Version::V1_2 {
+                Some(
+                    vk::PhysicalDeviceVulkan12Features::builder()
+                        .sampler_mirror_clamp_to_edge(
+                            features.contains(Features::SAMPLER_MIRROR_CLAMP_EDGE),
+                        )
+                        .draw_indirect_count(features.contains(Features::DRAW_INDIRECT_COUNT))
+                        .descriptor_indexing(
+                            features.intersects(Features::DESCRIPTOR_INDEXING_MASK),
+                        )
+                        .sampler_filter_minmax(true)
+                        .imageless_framebuffer(true)
+                        .build(),
+                )
+            } else {
+                None
+            },
+            descriptor_indexing: if enabled_extensions
```

Comment:
Note that this (and others) read from the enabled extensions list, rather than the features flags, since the "valid usage" for `VkDeviceCreateInfo` states that we shouldn't have these structs in the `pNext` chain if we didn't enable the extensions.